src/module-init/module-init.c

```diff
@@ -3,14 +3,159 @@ REDISMODULE_INIT_SYMBOLS();
+  // embedded newline in them
+  if (tolower(*value) == 'n') {
+    RedisModule_Log(ctx, "warning", "FATAL: aof-use-rdb-preamble required if AOF is used!");
+    rc = 0;
+  }
+  RedisModule_FreeCallReply(reply);
+  RedisModule_FreeThreadSafeContext(confCtx);
+  return rc;
+}
+
+static int initAsModule(RedisModuleCtx *ctx) {
+  // Check that redis supports thread safe context. RC3 or below doesn't
+  if (RedisModule_GetThreadSafeContext == NULL) {
+    RedisModule_Log(ctx, "warning",
+                    "***** FATAL: Incompatible version of redis 4.0 detected. *****\n"
+                    "\t\t\t\tPlease use Redis 4.0.0 or later from https://redis.io/download\n"
+                    "\t\t\t\tRedis will exit now!");
+    return REDISMODULE_ERR;
+  }
+
+  if (RediSearch_ExportCapi(ctx) != REDISMODULE_OK) {
+    RedisModule_Log(ctx, "warning", "Could not initialize low level api");
+  } else {
+    RedisModule_Log(ctx, "notice", "Low level api version %d initialized successfully",
+                    REDISEARCH_CAPI_VERSION);
+  }
+
+  if (RedisModule_GetContextFlags == NULL && RSGlobalConfig.concurrentMode) {
+    RedisModule_Log(ctx, "warning",
+                    "GetContextFlags unsupported (need Redis >= 4.0.6). Commands executed in "
+                    "MULTI or LUA will "
+                    "malfunction unless 'safe' functions are used or SAFEMODE is enabled.");
+  }
+
+  if (!validateAofSettings(ctx)) {
+    return REDISMODULE_ERR;
+  }
+
+  return REDISMODULE_OK;
+}
+
+static int initAsLibrary(RedisModuleCtx *ctx) {
+  // Ensure Redis symbols are initialized as well!.
+  // This is copy/pasted from redismodule.h
+  // We don't use RedisModule_Init, since this also changes the attributes
+  // of the provided ctx, which is probably owned by another module
+  RedisModule_GetApiFunctionType getapifuncptr = (RedisModule_GetApiFunctionType)((void **)ctx)[0];
+#define X(basename) getapifuncptr("RedisModule_" #basename, (void *)&RedisModule_##basename);
+  REDISMODULE_XAPI(X)
+#undef X
+}
+
+int RediSearch_Init(RedisModuleCtx *ctx, int mode) {
+#define DO_LOG(...)                               \
+  if (ctx && (mode == REDISEARCH_INIT_LIBRARY)) { \
+    RedisModule_Log(ctx, ##__VA_ARGS__);          \
+  }
+
+  // Print version string!
+  DO_LOG("notice", "RediSearch version %d.%d.%d (Git=%s)", REDISEARCH_VERSION_MAJOR,
+         REDISEARCH_VERSION_MINOR, REDISEARCH_VERSION_PATCH, RS_GetExtraVersion());
+
+  if (mode == REDISEARCH_INIT_MODULE && initAsModule(ctx) != REDISMODULE_OK) {
+    return REDISMODULE_ERR;
+  } else if (mode == REDISEARCH_INIT_LIBRARY && initAsLibrary(ctx) != REDISMODULE_OK) {
+    return REDISMODULE_ERR;
+  }
+
+  sds confstr = RSConfig_GetInfoString(&RSGlobalConfig);
+  DO_LOG("notice", confstr);
+  sdsfree(confstr);
+
+  // Init extension mechanism
+  Extensions_Init();
+
+  if (RSGlobalConfig.concurrentMode) {
+    ConcurrentSearch_ThreadPoolStart();
```

Comment:
Do we need it on library?

Reply:
I didn't want to complicate the code. The globals are de-initialized at module shutdown anyway, so I don't think it matters much..

Reply:
But it will open threads doesn't it?

Reply:
Probably the better way to solve this is to simply set concurrentMode to off

Reply:
So lets set it if loaded as static lib